dao/chat.go

```diff
@@ -2,8 +2,48 @@ package dao
 
 import (
 	"TUM-Live/model"
+	"errors"
+	"github.com/go-sql-driver/mysql"
 )
 
 func AddMessage(chat *model.Chat) error {
 	return DB.Save(chat).Error
 }
+
+//ToggleLike adds a like to a message from the user if it doesn't exist, or removes it if it does
+func ToggleLike(userID uint, chatID uint) error {
+	err := DB.Exec("INSERT INTO chat_user_likes (user_id, chat_id) VALUES (?, ?)", userID, chatID).Error
+	if err == nil {
+		return nil // like was added successfully
+	}
+	var mysqlErr *mysql.MySQLError
+	if errors.As(err, &mysqlErr) && mysqlErr.Number == 1062 { // 1062: duplicate entry -> message already liked -> remove
+		return DB.Exec("DELETE FROM chat_user_likes WHERE user_id = ? AND chat_id = ?", userID, chatID).Error
+	}
+	return err // some other error
+}
+
+// GetNumLikes returns the number of likes for a message
+func GetNumLikes(chatID uint) (int, error) {
+	var numLikes int
+	err := DB.Raw("SELECT COUNT(*) FROM chat_user_likes WHERE chat_id = ?", chatID).Scan(&numLikes).Error
```

Comment:
Maybe we can use https://gorm.io/docs/advanced_query.html#Count here?

Reply:
Definitely :+1: